src/protocolsupport/protocol/utils/spoofedata/SpoofedDataParser.java

```diff
@@ -9,17 +9,33 @@
 	private static final List<Function<String, SpoofedData>> parsers = new ArrayList<>();
 
 	static {
+		if (isPaperHandshakeEvent()) {
+			parsers.add(new PaperSpoofedDataParser());
+		}
 		parsers.add(new BungeeCordSpoofedDataParser());
 	}
 
 	public static SpoofedData tryParse(String data) {
-		return parsers.stream().map(parser -> {
+		for (Function<String, SpoofedData> parser : parsers) {
 			try {
-				return parser.apply(data);
+				SpoofedData result = parser.apply(data);
+				if (result != null) {
+					return result;
+				}
 			} catch (Exception e) {
-				return null;
+				// ignore
 			}
-		}).filter(res -> res != null).findAny().orElse(null);
+		}
+		return null;
+	}
+
+	private static boolean isPaperHandshakeEvent() {
```

Comment:
We might combine this with @MrPowerGamerBR's mcpe paper commit. Both PR's should maybe rely on paper detection at start and not searching class files in runtime to see if we're running paper or not.

Reply:
It's more than just "Paper" detection.

Older versions didn't provide the event.